File: PinetreeShop/PinetreeShop.Domain.Orders/OrderAggregate.cs
using PinetreeCQRS.Infrastructure;
using PinetreeShop.Domain.Orders.Commands;
using PinetreeShop.Domain.Orders.Events;
using PinetreeShop.Domain.Shared.Exceptions;
using PinetreeShop.Domain.Shared.Types;
using PinetreeShop.Domain.Tests.Order.Exceptions;
using System;
using System.Collections.Generic;
using System.Linq;

namespace PinetreeShop.Domain.Orders
{
    public class OrderAggregate : AggregateBase
    {
        public enum OrderState { Pending, ReadyForShipping, Shipped, Cancelled, Delivered };
        public OrderState State { get; set; }
        public IList<OrderLine> OrderLines { get; set; }
        public Guid BasketId { get; set; }
        public Guid ProcessId { get; set; }
        public Address ShippingAddress { get; set; }

        #region Constructors 

        public OrderAggregate()
        {
            RegisterEventHandler<OrderCreated>(Apply);
            RegisterEventHandler<OrderLineAdded>(Apply);
            RegisterEventHandler<OrderReadyForShipping>(Apply);
            RegisterEventHandler<OrderCancelled>(Apply);
            RegisterEventHandler<OrderShipped>(Apply);
            RegisterEventHandler<OrderDelivered>(Apply);
        }

        public OrderAggregate(CreateOrder cmd) : base()
        {
            var orderId = cmd.AggregateId;
            var basketId = cmd.BasketId;
            var shippingAddress = cmd.ShippingAddress;

            if (shippingAddress == null) throw new ParameterNullException(orderId, "shippingAddress");

            RaiseEvent(new OrderCreated(orderId, basketId, shippingAddress));
        }

#endregion

        #region Event handlers

        private void Apply(OrderReadyForShipping obj)
        {
            State = OrderState.ReadyForShipping;
        }

        private void Apply(OrderLineAdded obj)
        {
            OrderLines.Add(obj.OrderLine);
        }

        private void Apply(OrderCreated evt)
        {
            State = OrderState.Pending;
            OrderLines = new List<OrderLine>();

            AggregateId = evt.AggregateId;
            BasketId = evt.BasketId;
            ProcessId = evt.Metadata.ProcessId;
            ShippingAddress = evt.ShippingAddress;
        }

        private void Apply(OrderCancelled evt)
        {
            State = OrderState.Cancelled;
        }

        private void Apply(OrderShipped evt)
        {
            State = OrderState.Shipped;
        }

        private void Apply(OrderDelivered evt)
        {
            State = OrderState.Delivered;
        }

        #endregion

        #region Command handlers

        internal void PrepareForShipping(PrepareOrderForShipping cmd)
        {
            if (State != OrderState.Pending)
                throw new InvalidOrderStateException(cmd.AggregateId, $"State should be {OrderState.Pending} but is {State}");

            if (State == OrderState.ReadyForShipping)
                return;

            if(!OrderLines.Any())
                throw new InvalidOrderStateException(cmd.AggregateId, $"Order has no order lines");

            RaiseEvent(new OrderReadyForShipping(cmd.AggregateId));
        }

        internal void AddOrderLine(AddOrderLine cmd)
        {
            if(State != OrderState.Pending)
                throw new InvalidOrderStateException(cmd.AggregateId, $"State should be {OrderState.Pending} but is {State}");

            RaiseEvent(new OrderLineAdded(cmd.AggregateId, cmd.OrderLine));
        }

        internal void Cancel(CancelOrder cmd)
        {
            switch (State)
            {
                case OrderState.Shipped:
                case OrderState.Delivered:
                    throw new InvalidOrderStateException(cmd.AggregateId, $"State should be {OrderState.Pending} or {OrderState.ReadyForShipping} but is {State}");
                default:
                    RaiseEvent(new OrderCancelled(cmd.AggregateId));
                    break;
            }
        }

        internal void Ship(ShipOrder cmd)
        {
            if (State != OrderState.ReadyForShipping)
                throw new InvalidOrderStateException(cmd.AggregateId, $"State should be {OrderState.ReadyForShipping} but is {State}");

            RaiseEvent(new OrderShipped(cmd.AggregateId));
        }

        internal void Deliver(DeliverOrder cmd)
        {
            if (State != OrderState.Shipped) 
                throw new InvalidOrderStateException(cmd.AggregateId, $"State should be {OrderState.Shipped} but is {State}");

            RaiseEvent(new OrderDelivered(cmd.AggregateId));
        }

        internal static OrderAggregate Create(CreateOrder cmd)
        {
            return new OrderAggregate(cmd);
        }

        #endregion
    }
}

File: PinetreeShop/PinetreeShop.Domain.Orders/OrderAggregate.fs
module PinetreeShop.Domain.Orders.OrderAggregate

open System
open PinetreeCQRS.Infrastructure.Types
open PinetreeCQRS.Infrastructure.Commands
open PinetreeCQRS.Infrastructure.Events
open System
open Chessie.ErrorHandling
open PinetreeCQRS.Infrastructure.Validation

let orderCategory = Category "Order"
let orderQueueName = QueueName "Order"

type OrderError = 
    | ValidationError of string
    interface IError
    override e.ToString() = sprintf "%A" e

type OrderState = 
    | NotCreated
    | Pending
    | ReadyForShipping
    | Shipped
    | Delivered
    | Cancelled
    
type Command = 
    | Create of BasketId * ShippingAddress * OrderLine list
    | ReserveOrderLineProduct of ProductId
    | PrepareForShipping
    | Ship
    | Deliver
    | Cancel 
    interface ICommand

type Event = 
    | OrderCreated of BasketId * ShippingAddress * OrderLine list
    | OrderLineProductReserved of ProductId
    | OrderReadyForShipping
    | OrderShipped
    | OrderDelivered
    | OrderCancelled
    interface IEvent

module private Handlers = 
    type State = 
        { OrderState : OrderState
          OrderLines : (OrderLine * bool) list
          ShippingAddress : ShippingAddress }
        static member Zero = 
            { OrderState = OrderState.NotCreated
              OrderLines = []
              ShippingAddress = ShippingAddress "" }
    
    let markReserved (ol, res) productId =
        if (ol.ProductId = productId) then (ol, true)
        else (ol, res)
        

    let applyEvent state event = 
        match event with
        | OrderCreated(basketId, shippingAddress, orderLines) -> 
            { state with OrderState = OrderState.Pending
                         ShippingAddress = shippingAddress
                         OrderLines = List.map (fun o -> (o, false)) orderLines }
        | OrderLineProductReserved productId -> { state with OrderLines = List.map (fun ol -> markReserved ol productId) state.OrderLines }
        | OrderReadyForShipping -> { state with OrderState = OrderState.ReadyForShipping }
        | OrderShipped -> { state with OrderState = OrderState.Shipped }
        | OrderDelivered -> { state with OrderState = OrderState.Delivered }
        | OrderCancelled -> { state with OrderState = OrderState.Cancelled }
    
    module private Validate = 
        let inCase predicate error value = 
            if predicate value then  Bad [ ValidationError error :> IError ]
            else ok value
        
        module private Helpers = 
            let notEmptyShippingAddress (ShippingAddress sa) = 
                inCase (fun sa -> String.IsNullOrWhiteSpace(sa)) "Shipping Address cannot be empty" sa
            let isOrderState s os = 
                inCase (fun s -> s.OrderState <> os) (sprintf "Wrong Order state %A" s.OrderState) s
            let notOrderState s os = 
                inCase (fun s -> s.OrderState = os) (sprintf "Wrong Order state %A" s.OrderState) s
            let hasOrderLines ol = inCase (fun ol -> ol = []) "No order lines" ol
            let canCreate s = isOrderState s NotCreated
            let created s = notOrderState s NotCreated
        
        let canCreate (s, sa, ol) = Helpers.canCreate s <* Helpers.notEmptyShippingAddress sa <* Helpers.hasOrderLines ol
        let canMarkLineReserved s = Helpers.isOrderState s Pending
        let canPrepForShipping s = Helpers.isOrderState s Pending <* Helpers.hasOrderLines s.OrderLines
        let canShip s = Helpers.isOrderState s ReadyForShipping 
        let canDeliver s = Helpers.isOrderState s Shipped
        let canCancel s = Helpers.notOrderState s NotCreated <* Helpers.notOrderState s Cancelled <* Helpers.notOrderState s Delivered
    
    let executeCommand (state : State) command = 
        match command with
        | Create(basketId, shippingAddress, orderLines) -> 
            Validate.canCreate (state, shippingAddress, orderLines) <?> [ OrderCreated(basketId, shippingAddress, orderLines) ]
        | ReserveOrderLineProduct productId -> 
            let noOfOrders = state.OrderLines.Length
            let noOfReservedItems = List.filter (fun (ol, res) -> res = true) state.OrderLines |> List.length
            let events = 
                if (noOfOrders = noOfReservedItems + 1) then [ OrderLineProductReserved productId; OrderReadyForShipping ]
                else [ OrderLineProductReserved productId ]
            Validate.canMarkLineReserved state <?> events
        | PrepareForShipping -> Validate.canPrepForShipping (state) <?> [ OrderReadyForShipping ]
        | Ship -> Validate.canShip (state) <?> [ OrderShipped ]
        | Deliver -> Validate.canDeliver (state) <?> [ OrderDelivered ]
        | Cancel -> Validate.canCancel (state) <?> [ OrderCancelled ]

let makeOrderCommandHandler = 
    makeCommandHandler { Zero = Handlers.State.Zero
                         ApplyEvent = Handlers.applyEvent
                         ExecuteCommand = Handlers.executeCommand }
